controller/fsm.py

```diff
@@ -0,0 +1,63 @@
+"""Module that implements `drunc` finite state machine."""
+
+from statemachine import State, StateMachine
+
+
+class DruncFSM(StateMachine):
+    """Finite state machine for the `drunc` controller."""
+
+    # The states the controller can be in
+    none = State("None", initial=True)
+    initial = State("Initial")
+    configured = State("Configured")
+    ready = State("Ready")
+    running = State("Running")
+    dataflow_drained = State("Dataflow_Drained")
+    triggered_sources_stopped = State("Triggered_Sources_Stopped")
+
+    # The transitions between the states
+    boot = none.to(initial)
+    terminate = initial.to(none)
+    conf = initial.to(configured)
+    scrap = configured.to(initial)
+    start = configured.to(ready)
+    enable_triggers = ready.to(running)
+    disable_triggers = running.to(ready)
+    drained_dataflow = ready.to(dataflow_drained)
+    stop_triggered_sources = dataflow_drained.to(triggered_sources_stopped)
+    stop = triggered_sources_stopped.to(configured)
+
+    def to_dict(self) -> dict[str, list[dict[str, str]]]:
+        """Return the FSM states and events as a dictionary.
+
+        The states will be the keys and the valid events for each state a list of
+        values.
+
+        Returns:
+            dict[str, list[str]]: The states and events as a dictionary.
+        """
+        states = {
+            state.name: [
+                {"event": transition.event, "target": transition.target.name}
+                for transition in state.transitions
+            ]
+            for state in self.states
+        }
+        return {state: states[state] for state in self.ordered_states}
+
+    @property
+    def ordered_states(self) -> list[str]:
+        """Return the order of the states.
+
+        Returns:
+            list[str]: The order of the states.
+        """
+        return [
+            "None",
+            "Initial",
+            "Configured",
+            "Ready",
+            "Running",
+            "Dataflow_Drained",
+            "Triggered_Sources_Stopped",
+        ]
```

Comment:
This is a good stand in for the controller interface but we shouldn't need it once we have that interface working otherwise we risk duplicating the state of the controller.

Reply:
Well, yes and no. Based on the conversations with Pierre, out of the controller you can get:
- The current state
- The transitions allowed from this state
- All the possible transitions

But you cannot get:
- All possible states
- To what state each transition leads you

So it is not possible to draw a map of the FSM out of the information facilitated by the Controller - we need to hardcode it somehow.

Having said that, I totally agree that the Controller should be the single source of truth of the current state, so some of the features of this structure are either redundant or will not be used. You can be seen more as an aid to the representation of the machine, if you want, but the state is not really hold by it, but by the Controller.

Reply:
I've simplified things massively using simple mappings. You were definitely right there.